content/assignments/memory-allocator/README.md

```diff
@@ -1 +1,273 @@
+
+   Chunks of memory smaller than `MMAP_THRESHOLD` are allocated with `brk()`.
+   Bigger chunks are allocated using `mmap()`.
+   The memory is uninitialized.
+
+   - Passing `0` as `size` will return `NULL`
+
+1. `void *os_calloc(size_t nmemb, size_t size)`
+
+   Allocates memory for an array of `nmemb` elements of `size` bytes each and returns a pointer to the allocated memory.
+
+   Chunks of memory smaller than [`page_size`](https://man7.org/linux/man-pages/man2/getpagesize.2.html) are allocated with `brk()`.
+   Bigger chunks are allocated using `mmap()`.
+   The memory is set to zero.
+
+   - Passing `0` as `nmemb` or `size` will return `NULL`
+
+1. `void *os_realloc(void *ptr, size_t size)`
+
+   Changes the size of the memory block pointed to by `ptr` to `size` bytes.
+   If the size is smaller than the previously allocated size, the memory block will be truncated.
+
+   If `ptr` points to a block on heap, `os_realloc()` will first try to expand the block, rather than moving it.
+   Otherwise, the block will be reallocated and its contents copied.
+
+   When attempting to expand a block followed by multiple free blocks, `os_realloc()` will coalesce them one at a time and verify the condition for each.
+   Blocks will remain coalesced even if the resulting block will not be big enough for the new size.
+
+   Calling `os_realloc()` on a block that has `STATUS_FREE` should return `NULL`.
+   This is a measure to prevent undefined behavior and make the implementation robust, it should not be considered a valid use case of `os_realloc()`.
+
+   - Passing `NULL` as `ptr` will have the same effect as `os_malloc(size)`
+   - Passing `0` as `size` will have the same effect as `os_free(ptr)`
+
+1. `void os_free(void *ptr)`
+
+   Frees memory previously allocated by `os_malloc()`, `os_calloc()` or `os_realloc()`.
+
+   `os_free()` will not return memory from the heap to the OS by calling `brk()`, but rather mark it as free and reuse it in future allocations.
+   In the case of mapped memory blocks, `os_free()` will call `munmap()`.
+
+1. General
+
+   - Allocations that increase the heap size will only expand the last block if it is free.
+   - You must check the error code returned by every syscall.
+   You can use the `DIE()` macro for this.
+
+## Implementation
+
+An efficient implementation must keep data aligned, keep track of memory blocks and reuse freed blocks.
+This can be further improved by reducing the number of syscalls and block operations.
+
+### [Memory Alignment](https://stackoverflow.com/a/381368)
+
+Allocated memory should be aligned (i.e. all addresses are multiple of a given size).
+This is a space-time tradeoff because memory blocks are padded so each can be read in one transaction.
+It also allows for atomicity when interacting with a block of memory.
+
+All memory allocations should be aligned to **8 bytes** as required by 64 bit systems.
+
+### Block Reuse
+
+#### `struct block_meta`
```

Comment:
I don't know if this is the best option.
It doesn't look very good imo, but I don't have another idea.

Reply:
Are you talking about the 4 `#` or about using the structure as a section name?